File: CalculCpp/Calcul.cpp
#include <iostream>
#include <math.h>
#include <stdio.h>
#include <string>
#include <thread>

double doCalc(double x) { return x * x; }

void calculPortion(const double min, const double max, const unsigned long segments)
{
	double resultat, tmp;
	double x = min;

	double pas = (max - min) / (double) segments;

	for (unsigned long i = 0; i < segments; i++) 
	{
		resultat += doCalc(x) * pas;
		x += pas;
	}

	//printf("De %e à %e : %f\n", min, max, resultat);
}

int main(int argc, char *argv[])
{
	
	// verify command line arguments
	unsigned long nbSegments, nbCores;
	double        min, max;
	
	for (int i = 0 ; i < argc ; i++)
	{
		printf("%s\n", argv[i]);
	}

	if(argc == 5)
	{
		nbCores    = std::stol(argv[1]);
		min        = std::stod(argv[2]);
		max        = std::stod(argv[3]);
		nbSegments = std::stol(argv[4]);
	}
	else
	{
		printf("Usage : %s <nbCores> <min> <max> <nbSegments>\n", argv[0]);
		exit(1);
	}

	if(min > max) 
	{
		printf("Wrong input, maximum is inferior to minimum...");
		exit(2);
	}

	if (nbCores < 0) {
		printf("Impossible to use less than 1 core...");
		exit(3);
	}

	double increment = (max - min) / (double) nbCores;
	max = min + increment;

	// on crée autant de thread fils que de cœurs sur la machine
	std::thread tabThreads[nbCores];

	for(unsigned int i = 0; i < nbCores; i++)
	{
		// creation of child threads for computing
		// store it in an array for easier access
		tabThreads[i] = std::thread(calculPortion, min, max, nbSegments);

		min += increment;
		max += increment;
	}

	for(unsigned int i = 0 ; i < nbCores; i++)
	{
		tabThreads[i].join();
	}
}

File: CalculGo/Calcul.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

// for computing a mathematic function
func doCalc(x float64) float64 {

	return x * x
}

// compute area in the given range
func calcSegment(min float64, max float64, segments int, chResultat chan string) {

	var resultat float64

	var pas = (max - min) / float64(segments)

	var x = min

	for i := 0; i < segments; i++ {
		resultat += doCalc(x) * pas
		x += pas
	}
	
	// convert number to strings
	// strMin := strconv.FormatFloat(min,      'E', -1, 64)
	// strMax := strconv.FormatFloat(max,      'E', -1, 64)
	// strRes := strconv.FormatFloat(resultat, 'E', -1, 64)
	
	//chResultat <- "From " + strMin + " to " + strMax + " = " + strRes
	
	chResultat <- "done"
}

func main() {

	var min, max float64
	var nbSegments, nbCores int64
	
	// verify command line arguments
	if len(os.Args) == 5 {

		nbCores, _    = strconv.ParseInt(os.Args[1], 10, 64)
		min, _        = strconv.ParseFloat(os.Args[2], 64) 
		max, _        = strconv.ParseFloat(os.Args[3], 64)
		nbSegments, _ = strconv.ParseInt(os.Args[4], 10, 64)
		

	} else {
		fmt.Printf("Usage : %s <nbCores> <min> <max> <nbSegments>\n", os.Args[0])
		os.Exit(1)
	}

	if min > max {
		fmt.Println("Wrong input, maximum is inferior to minimum...")
		os.Exit(2)
	}
	
	if nbCores < 0 {
		fmt.Println("Impossible to use less than 1 core...")
		os.Exit(3)
	}

	// compute the increment, to distribute the same part to all the threads
	increment  := (max - min) / float64(nbCores)
	chResultat := make(chan string)
	
	max = min + increment

	for i := 0 ; i < int(nbCores) ; i++ {
		
		// creation of child threads for computing
		go calcSegment(min, max, int(nbSegments/nbCores), chResultat)

		// shifting the range for the next process
		min += increment
		max += increment
	}

	// we fetch and display data
	for i := 0; i < int(nbCores) ; i++ {
		<-chResultat
	}
}